### PR TITLE
fix: Add columns for user metadata in the tables

### PR DESCRIPTION
**Pull request by achals**
**What this PR does / why we need it**:

Additionally, this PR refactors feast to use the registry abstract base class wherever possible.

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

File: sdk/python/feast/feature_store.py [hunk 1]
```diff
@@ -142,8 +141,9 @@ def __init__(
         if registry_config.registry_type == "sql":
             self._registry = SqlRegistry(registry_config, None)
         else:
-            self._registry = Registry(registry_config, repo_path=self.repo_path)
-            self._registry._initialize_registry()
+            r = Registry(registry_config, repo_path=self.repo_path)
```
Comment (kevjumba) [inline sdk/python/feast/feature_store.py:144]: why is this change necessary?

File: sdk/python/feast/infra/registry_stores/sql.py [hunk 2]
```diff
@@ -300,13 +353,14 @@ def apply_data_source(
...
         self, feature_view: BaseFeatureView, project: str, commit: bool = True
     ):
-        # TODO(achals): Stream feature views need to be supported.
         if isinstance(feature_view, FeatureView):
             fv_table = feature_views
         elif isinstance(feature_view, OnDemandFeatureView):
             fv_table = on_demand_feature_views
         elif isinstance(feature_view, RequestFeatureView):
             fv_table = request_feature_views
+        elif isinstance(feature_view, StreamFeatureView):
```
Comment (kevjumba) [inline sdk/python/feast/infra/registry_stores/sql.py:362]: this actually doesn't work since streamfeatureview is actually an instance of featureview(it inherits feature view) Either streamfeatureview has to go first or the first isinstance has to assert that it's not a batch or streamfeatureview.

File: sdk/python/feast/infra/registry_stores/sql.py [hunk 3]
```diff
@@ -406,8 +460,100 @@ def apply_materialization(
...
+        feature_view: BaseFeatureView,
+        metadata_bytes: Optional[bytes],
+    ):
+        if isinstance(feature_view, FeatureView):
+            table = feature_views
+        elif isinstance(feature_view, OnDemandFeatureView):
+            table = on_demand_feature_views
+        elif isinstance(feature_view, RequestFeatureView):
+            table = request_feature_views
+        elif isinstance(feature_view, StreamFeatureView):
```
Comment (kevjumba) [inline sdk/python/feast/infra/registry_stores/sql.py:489]: same here

File: sdk/python/feast/infra/registry_stores/sql.py [hunk 4]
```diff
@@ -406,8 +460,100 @@ def apply_materialization(
...
+    def get_user_metadata(
+        self, project: str, feature_view: BaseFeatureView
+    ) -> Optional[bytes]:
+        if isinstance(feature_view, FeatureView):
+            table = feature_views
+        elif isinstance(feature_view, OnDemandFeatureView):
+            table = on_demand_feature_views
+        elif isinstance(feature_view, RequestFeatureView):
+            table = request_feature_views
+        elif isinstance(feature_view, StreamFeatureView):
```
Comment (kevjumba) [inline sdk/python/feast/infra/registry_stores/sql.py:523]: and here

File: sdk/python/feast/registry.py [hunk 5]
```diff
@@ -337,9 +337,22 @@ def delete_feature_view(self, name: str, project: str, commit: bool = True):
...
-    #     ...
+    @abstractmethod
+    def get_stream_feature_view(
+        self, name: str, project: str, allow_cache: bool = False
+    ):
+        """
+        Retrieves a stream feature view.
+
+        Args:
+            name: Name of feature view
```
Comment (kevjumba) [inline sdk/python/feast/registry.py:348]: stream feature view

File: sdk/python/feast/registry.py [hunk 6]
```diff
@@ -423,10 +436,20 @@ def list_feature_views(
...
-    # @abstractmethod
-    # def get_request_feature_view(self, name: str, project: str):
-    #     ...
+    @abstractmethod
+    def get_request_feature_view(self, name: str, project: str) -> RequestFeatureView:
+        """
+        Retrieves a request feature view.
+
+        Args:
+            name: Name of feature view
```
Comment (kevjumba) [inline sdk/python/feast/registry.py:445]: request feature view

**Review by felixwang9817 — COMMENTED** [review]
lgtm barring Kevin's comments

File: sdk/python/feast/feature_store.py (hunk 1 above)
Comment (achals) [reply sdk/python/feast/feature_store.py:144]: `self._registry` has type `BaseRegistry` which doesn't have an `_initialize_registry` method.

File: sdk/python/feast/infra/registry_stores/sql.py (hunk 2 above)
Comment (achals) [reply sdk/python/feast/infra/registry_stores/sql.py:362]: Hmmm why does StreamFeatureView subclass FeatureView?

File: sdk/python/feast/feature_store.py (hunk 1 above)
Comment (achals) [reply sdk/python/feast/feature_store.py:144]: `self._registry` has type `BaseRegistry` which doesn't have an `_initialize_registry` method.

File: sdk/python/feast/infra/registry_stores/sql.py [hunk 7]
```diff
@@ -87,6 +93,16 @@
     Column("feature_view_name", String(50), primary_key=True),
     Column("last_updated_timestamp", BigInteger, nullable=False),
     Column("feature_view_proto", LargeBinary, nullable=False),
+    Column("user_metadata", LargeBinary, nullable=True),
+)
+
+streaming_feature_views = Table(
+    "streaming_feature_views",
```
[the remark this answers is not in the document]
Comment (achals) [reply sdk/python/feast/infra/registry_stores/sql.py]: done

File: sdk/python/feast/registry.py (hunk 6 above)
Comment (achals) [reply sdk/python/feast/registry.py]: done

File: sdk/python/feast/registry.py (hunk 5 above)
Comment (achals) [reply sdk/python/feast/registry.py]: done

File: sdk/python/feast/infra/registry_stores/sql.py (hunk 4 above)
Comment (achals) [reply sdk/python/feast/infra/registry_stores/sql.py]: done

File: sdk/python/feast/infra/registry_stores/sql.py (hunk 3 above)
Comment (achals) [reply sdk/python/feast/infra/registry_stores/sql.py]: done

File: sdk/python/feast/infra/registry_stores/sql.py (hunk 2 above)
Comment (achals) [reply sdk/python/feast/infra/registry_stores/sql.py]: done
